geo/src/algorithm/sweep/mod.rs

```diff
@@ -1,29 +1,364 @@
+///   and check all INSERT events that start before that DELETE x-coordinate.
+///
+/// By maintaining separate vecs for INSERT and DELETE events and using the segment index
+/// for correlation, we can determine which segments are active at any point
+/// _without_ maintaining an "active segment" tree or using a HashMap.
+#[derive(Debug, Clone)]
+struct SweepLineEvent<T: GeoFloat> {
+    x_value: T,
+    event_type: EventType,
+    interval: SweepLineInterval<T>,
+}
+
+impl<T: GeoFloat> SweepLineEvent<T> {
+    fn new(x: T, event_type: EventType, interval: SweepLineInterval<T>) -> Self {
+        Self {
+            x_value: x,
+            event_type,
+            interval,
+        }
+    }
+}
+
+impl<T: GeoFloat> PartialEq for SweepLineEvent<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.x_value.total_cmp(&other.x_value) == Ordering::Equal
+            && self.event_type == other.event_type
+    }
+}
+
+impl<T: GeoFloat> Eq for SweepLineEvent<T> {}
+
+impl<T: GeoFloat> PartialOrd for SweepLineEvent<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: GeoFloat> Ord for SweepLineEvent<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Use total_cmp for more robust floating point comparison
+        // This is crucial for grid patterns where many x-values are exactly equal
+        match self.x_value.total_cmp(&other.x_value) {
+            Ordering::Equal => {
+                // When x values are equal, prioritize INSERT events before DELETE events
+                // This ensures proper handling of overlapping segments
+                match (self.event_type, other.event_type) {
+                    (EventType::Insert, EventType::Delete) => Ordering::Less,
+                    (EventType::Delete, EventType::Insert) => Ordering::Greater,
+                    _ => Ordering::Equal,
+                }
+            }
+            ordering => ordering,
+        }
+    }
+}
+
+struct SweepLineIndex<T: GeoFloat> {
+    insert_events: Vec<SweepLineEvent<T>>,
+    delete_events: Vec<SweepLineEvent<T>>,
+}
+
+impl<T: GeoFloat> SweepLineIndex<T> {
+    fn new(num_segments: usize) -> Self {
+        Self {
+            insert_events: Vec::with_capacity(num_segments),
+            delete_events: Vec::with_capacity(num_segments),
+        }
+    }
+
+    fn add(&mut self, interval: SweepLineInterval<T>) {
+        let x_min = interval.min;
+        let x_max = interval.max;
+
+        // Create INSERT event
+        let insert_event = SweepLineEvent::new(x_min, EventType::Insert, interval);
+        self.insert_events.push(insert_event);
+
+        // Create DELETE event
+        let delete_event = SweepLineEvent::new(x_max, EventType::Delete, interval);
+        self.delete_events.push(delete_event);
+    }
+
+    fn compute_intersections<F>(&mut self, mut intersection_action: F)
```

Comment:
take it or leave it: This name confused me a bit. 

I assumed `intersection_action` was called when there was an intersection, e.g. synonymous with `on_intersection` or `handle_intersection`, but it's actually called when there is an *overlap* and it is responsible for determining when there is an intersection. 

Better names might be: `check_for_intersection` or (similar to JTS) `overlap_action`